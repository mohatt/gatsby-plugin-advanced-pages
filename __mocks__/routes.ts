import type { SerializedRoute } from '@/node'

export default <SerializedRoute[]>[
  {
    name: 'home',
    path: '/',
    scopes: {},
  },
  {
    name: 'about',
    path: '/about',
    scopes: {},
  },
  {
    name: 'blog',
    path: '/blog',
    scopes: {
      pagination: '/blog/page/:page',
    },
  },
  {
    name: 'blog.author',
    path: '/blog/author/:author',
    scopes: {
      pagination: '/blog/author/:author/page/:page',
    },
  },
  {
    name: 'blog.tag',
    path: '/blog/tag/:tag',
    scopes: {
      pagination: '/blog/tag/:tag/page/:page',
    },
  },
  {
    name: 'blog.post',
    path: '/blog/post/:post',
    scopes: {},
  },
]
