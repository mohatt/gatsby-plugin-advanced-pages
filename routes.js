module.exports = {
  "projects": "/portafolio",
  "projects.paginated": "/portafolio/page/:page",
  "projects.skill": "/portafolio/:skill",
  "projects.skill.paginated": "/portafolio/:skill/page/:page"
}
