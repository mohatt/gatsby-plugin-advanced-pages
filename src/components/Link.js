import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { string, object } from 'prop-types'
import { routeExists, generatePath } from '../api'

const Link = ({ to, params, scope, ...props }) => {
  if (to && routeExists(to)) {
    try {
      to = generatePath(to, params, scope, true)
    } catch (e) {
      console.error(
        'Warning: Invalid route params supplied to `Link` component: ' + e.message
      )
    }
  }

  return (
    <GatsbyLink to={to} {...props} />
  )
}

Link.propTypes = {
  to: string.isRequired,
  params: object,
  scope: string
}

export default Link
