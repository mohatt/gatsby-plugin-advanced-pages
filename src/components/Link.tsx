import React, { ReactNode } from 'react'
import { Link as GatsbyLink, GatsbyLinkProps } from 'gatsby'
import { string, object } from 'prop-types'
import { routeExists, generatePath } from '../api'
import type { RouteScope } from '../gatsby/types'

export interface LinkProps<TState> extends GatsbyLinkProps<TState> {
  to: string
  params?: Record<string, any>
  scope?: RouteScope
}

export const Link = <TState,>({ to, params, scope, ...props }: LinkProps<TState>): ReactNode => {
  if (to != null && routeExists(to)) {
    try {
      to = generatePath(to, params, scope, true)
    } catch (e) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(
          `gatsby-plugin-advanced-pages: Unable to generate path for route "${to}": ${e.message}`,
        )
      }
      to = '/404.html'
    }
  }

  return <GatsbyLink to={to} {...(props as any)} />
}

Link.propTypes = {
  to: string.isRequired,
  params: object,
  scope: string,
}

export default Link
