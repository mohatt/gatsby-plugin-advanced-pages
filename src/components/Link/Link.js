import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { string, object } from 'prop-types'
import { generatePath } from '../../api'

const Link = ({ to, params, scope, ...props }) => {
  let href
  if (to !== undefined) {
    try {
      href = generatePath(to, params, scope)
    } catch (e) {
      if (process.env.NODE_ENV !== `production`) {
        console.error(
          'Warning: Invalid route props supplied to `Link` component: ' + e.message
        )
      }
    }
  }

  return (
    <GatsbyLink to={href} {...props} />
  )
}

Link.propTypes = {
  to: string.isRequired,
  params: object,
  scope: string
}

export default Link
