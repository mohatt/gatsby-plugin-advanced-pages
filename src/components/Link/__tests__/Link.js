import React from 'react'
import TestRenderer from 'react-test-renderer'
import Link from '../Link'
import routes from '../../../../test/__fixtures__/routes'

// Create a virtual mock for routes.js
jest.doMock('../../../routes', () => routes, { virtual: true })

describe(`<Link />`, () => {
  const render = (props = {}) => {
    return TestRenderer.create(<Link {...props} />).toJSON()
  }

  it(`should render correctly`, () => {
    expect(render({ to: 'home' })).toMatchSnapshot()
  })

  it(`should render correctly with route params`, () => {
    expect(render({
      to: 'blog.post',
      params: {
        post: 'hello-world'
      }
    })).toMatchSnapshot()
  })

  it(`should render correctly with route params and scope`, () => {
    expect(render({
      to: 'blog.tag',
      scope: 'pagination',
      params: {
        tag: 'news',
        page: 5
      }
    })).toMatchSnapshot()
  })

  it(`should log console error when called with invalid props`, () => {
    const spy = jest.spyOn(console, 'error').mockImplementation()
    render()
    expect(spy.mock.calls).toMatchSnapshot()
    spy.mockClear()
    render({ to: 'blog.post' })
    expect(spy.mock.calls).toMatchSnapshot()
    spy.mockClear()
    render({
      to: 'blog.post',
      params: {
        invalid: 'whatever'
      }
    })
    expect(spy.mock.calls).toMatchSnapshot()
    spy.mockRestore()
  })

  it(`should render correctly when called with a non-existent route`, () => {
    expect(render({ to: 'invalid-route' })).toMatchSnapshot()
    expect(render({ to: 'some/path' })).toMatchSnapshot()
  })
})
