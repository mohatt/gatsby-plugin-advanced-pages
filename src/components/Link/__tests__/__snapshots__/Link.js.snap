// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should log console error when called with invalid props 1`] = `
Array [
  Array [
    "Warning: Failed prop type: The prop \`to\` is marked as required in \`Link\`, but its value is \`undefined\`.
    in Link (at Link.js:11)",
  ],
]
`;

exports[`<Link /> should log console error when called with invalid props 2`] = `
Array [
  Array [
    "Warning: Invalid route params supplied to \`Link\` component: Error generating a path for route '/blog/post/:post' with params '{}': Expected \\"post\\" to be a string",
  ],
]
`;

exports[`<Link /> should log console error when called with invalid props 3`] = `
Array [
  Array [
    "Warning: Invalid route params supplied to \`Link\` component: Error generating a path for route '/blog/post/:post' with params '{\\"invalid\\":\\"whatever\\"}': Expected \\"post\\" to be a string",
  ],
]
`;

exports[`<Link /> should render correctly 1`] = `
<a
  href="/site/"
/>
`;

exports[`<Link /> should render correctly when called with a non-existent route 1`] = `
<a
  href="/site/invalid-route"
/>
`;

exports[`<Link /> should render correctly when called with a non-existent route 2`] = `
<a
  href="/site/some/path"
/>
`;

exports[`<Link /> should render correctly with route params 1`] = `
<a
  href="/site/blog/post/hello-world"
/>
`;

exports[`<Link /> should render correctly with route params and scope 1`] = `
<a
  href="/site/blog/tag/news/page/5"
/>
`;
